[Frontend]: Refactor PageSpeed monitor Create & Configure components

Jesulayomy opened the pull request:
## FE PageSpeed Monitor Create and Config Merger

- Refactored the Create and Configure components to use a single Setup component and reused most of the code
- This PR Refactors the PageSpeed Monitor

## Write your issue number after "Fixes "

Fixes #1304

## Please ensure all items are checked off before requesting a review. "Checked off" means you need to add an "x" character between brackets so they turn into checkmarks.

- [x] (Do not skip this or your PR will be closed) I deployed the application locally.
- [x] (Do not skip this or your PR will be closed) I have performed a self-review and testing of my code.
- [x] I have included the issue # in the PR.
- [x] I have added i18n support to visible strings (instead of `<div>Add</div>`, use): 
```Javascript
const { t } = useTranslation();
<div>{t('add')}</div>
```
- [x] I have **not** included any files that are not related to my pull request, including package-lock and package-json if dependencies have not changed
- [x] I didn't use any hardcoded values (otherwise it will not scale, and will make it difficult to maintain consistency across the application).
- [x] I made sure font sizes, color choices etc are all referenced from the theme. I don't have any hardcoded dimensions.
- [x] My PR is granular and targeted to one specific feature.
- [x] I ran `npm run format` in server and client directories, which automatically formats your code.
- [x] I took a screenshot or a video and attached to this PR if there is a UI change. (No visible UI changes)

Jesulayomy commented on client/src/Pages/PageSpeed/Create/index.jsx:
```diff
@@ -40,60 +45,98 @@ const SELECT_VALUES = [
-		interval: 3,
-	});
+const PageSpeedSetup = ({ configure }) => {
+	const isConfigure = configure;
```
I could take it out and cleanup the routes using something like this is more direct:
```js
const { monitorId } = useParams();
const isCreate = !!monitorId;
```

Jesulayomy commented on client/src/Pages/PageSpeed/Create/index.jsx:
```diff
@@ -40,60 +45,98 @@ const SELECT_VALUES = [
+					newErrors[err.path[0]] = err.message;
+				});
+				setErrors(newErrors);
+				createToast({ body: "Please check the form for errors." });
```
Do I also need to modify every other localization file to add the field there?

ajhollid replied:
> Do I also need to modify every other localization file to add the field there?

No this is not necessary, just the `en.json` for the baseline, localization folks will handle the other files.

ajhollid commented on client/src/Pages/PageSpeed/Create/index.jsx:
```diff
@@ -40,60 +45,98 @@ const SELECT_VALUES = [
-		};
+		if (isConfigure) {
+			await updateMonitor({ monitor, redirect: "/pagespeed" });
+		} else {
```
Validation should be done when submitting as well, it is the final gate keeper before data is sent off.

There's no guarantee that the behavior of `onChange` will stay as it is, but `onSubmit` will always be the step before submission so we want validation there always.

ajhollid commented on client/src/Pages/PageSpeed/Create/index.jsx:
```diff
@@ -1,99 +1,141 @@
-	});
+const PageSpeedSetup = () => {
+	const { monitorId } = useParams();
+	const isConfigure = !!monitorId;
```
Let's use the `isCreate` convention as in other pages please.

Jesulayomy replied:
Now using:

```js
const { monitorId } = useParams();
const isCreate = typeof monitorId === "undefined";
```

The other parts of the code will be adjusted accordingly to replace isConfigure variable

ajhollid commented on client/src/Pages/PageSpeed/Create/index.jsx:
```diff
@@ -1,99 +1,141 @@
+	const [updateTrigger, setUpdateTrigger] = useState(false);
+	const [isDeleting, setIsDeleting] = useState(false);
+	const [isUpdating, setIsUpdating] = useState(false);
+	const [isPausing, setIsPausing] = useState(false);
```
There's no need to track loading state here, I believe all of these hooks supply an `isLoading` state.

Jesulayomy replied:
Removed, `updateTrigger` is used when fetching the monitor based on an id, so it's kept in.

ajhollid reviewed (changes requested):
Thanks for making requested changes here, it's looking good.

We would like to use `isCreate` as a convention rather than `isConfigure`.

I noticed that you're tracking loading state locally within the component, this is not necessary as the hooks provide a loading state, please use those states.

Thanks for your contribution!

Jesulayomy commented on client/src/Pages/PageSpeed/Create/index.jsx:
```diff
@@ -1,36 +1,41 @@
-const SELECT_VALUES = [
+// Constants
+const MS_PER_MINUTE = 60000;
+const FREQUENCIES = [
```
Yes, added `time.threeMinutes, ...` to the en.json

Jesulayomy commented:
Made another [commit](https://github.com/bluewave-labs/Checkmate/pull/2588/commits/eaa85bb9c46aa61527a43ff2f6f45e1135102bb8) for the recent comments
@ajhollid

ajhollid reviewed (changes requested):
It's getting there :+1:  A coupel of issues I see at the moment

1. Unused classnames, these should be removed
2. Conditional calling of a hook, this should be avoided and conditional early returns added to the hook itself
3.  Loading states need to be used to set the `loading` prop of buttons so they show loading state and are disabled when an operation is in place.  If you are doing a netwokr operation all buttons that allow network operations to take place should be set to `loading`

Thanks for revisions!

Jesulayomy commented on client/src/Pages/PageSpeed/Create/index.jsx:
```diff
@@ -1,99 +1,154 @@
 	const [createMonitor, isCreating] = useCreateMonitor();
+	const [isLoading] = isCreate
+		? [false]
+		: useFetchMonitorById({ monitorId, setMonitor, updateTrigger });
```
I'll place a check for undefined monitorId in the hook itself so it doesn't try to fetch the monitor when it does not exist.
```js
if (monitorId !== undefined) {
// Fetch monitor
...
} else {
  // set loading to false before return
  setIsLoading(false);
}
```

Br0wnHammer commented on client/src/Pages/PageSpeed/Create/index.jsx:
```diff
@@ -103,210 +155,361 @@ const CreatePageSpeed = () => {
+								type="submit"
+								variant="contained"
+								color="accent"
+								loading={isLoading || isCreating || isDeleting || isUpdating || isPausing}
```
Same here. The logic can be reused.

Br0wnHammer reviewed (changes requested):
Looks good for the most part. Please review my comments. Some modifications are required.

Jesulayomy commented on client/src/Pages/PageSpeed/Create/index.jsx:
```diff
@@ -103,210 +155,361 @@ const CreatePageSpeed = () => {
+												"&:before": {
+													position: "absolute",
+													content: `""`,
+													width: 4,
```
I see the styling is for the circle that appears next to the Editing text. Do you think replacing this with a • would be reasonable?

Jesulayomy commented:
### Latest changes ♻️ 

- [x] Moved loading checks into a single `isBusy` variable
- [x] Replaced hard coded spacing with theme spacing
- [x] Moved skeleton layout render outside of main render

Br0wnHammer reviewed (approved):
LGTM! Thanks for your contributions 🚀

Br0wnHammer commented:
@Jesulayomy There are some format checks failing. Can you take care of that. Make sure to run `npm run format`.

Jesulayomy commented:
@Br0wnHammer Thanks for noticing that, I've fixed it now. 🚀
